File: hw3/yandex/b.js
const readline = require('readline');
const rl = readline.createInterface({ input: process.stdin });

rl.on('line', onInputLine)

let lines = 0;
let N;
let M;
let points = [];

function onInputLine(line) {
    let nums = line.split(' ').map(Number);

    if (lines === 0) {
        N = nums[0];
        M = nums[1];
    }
    else if (lines <= N) {
        points.push(
            { x: Math.min(nums[0], nums[1]), t: 0 },
            { x: Math.max(nums[0], nums[1]), t: 2 }
        );
    }
    else if (lines === N + 1) {
        points.push(...nums.map((x, i) => ({ x, t: 1, i })));
        pointsAndLines(N, M, points);
    }

    lines++;
}

function quickSort_tp(A) {
    function partition(left, right) {
        let pivot = A[parseInt((right + left) / 2)];
        let l = left;
        let r = right;

        while (l <= r) {
            while (A[l].x === pivot.x ? A[l].t < pivot.t : A[l].x < pivot.x) l++;
            while (A[r].x === pivot.x ? A[r].t > pivot.t : A[r].x > pivot.x) r--;

            if (l <= r) {
                [A[l], A[r]] = [A[r], A[l]];
                l++;
                r--;
            }
        }

        return l;
    }

    function qs(left, right) {
        while (right > left) {
            let p = partition(left, right);
            qs(p, right);
            right = p - 1;
        }
    }

    return qs(0, A.length - 1);
}

function pointsAndLines(N, M, points) {
    quickSort_tp(points);

    let count = 0;
    let result = Array.from({ length: M }, () => '0');
    for (let i = 0; i < points.length; i++) {
        let point = points[i];
        if (point.t === 1) result[point.i] = count.toString();
        if (point.t === 0) count += 1;
        if (point.t === 2) count -= 1;
    }

    console.log(result.join(' '));
}

File: hw3/yandex/b.py
import tracemalloc

tracemalloc.start()

def quickSort(
    A,
    compare=lambda a, pivot, pointer: a < pivot if pointer == 0 else a > pivot,
    left=None,
    right=None
):
    if left == None:
        left = 0
    if right == None:
        right = len(A) - 1
    if (left >= right):
        return A

    pivot = A[(right + left) // 2]
    l = left
    r = right

    while (l <= r):
        while (compare(A[l], pivot, 0)):
            l += 1
        while (compare(A[r], pivot, 1)):
            r -= 1

        if (l <= r):
            A[l], A[r] = A[r], A[l]
            l += 1
            r -= 1

    quickSort(A, compare, left, l - 1)
    quickSort(A, compare, l, right)

    return A

def compare(a,b,p):
    if (p ==0):
        return a[1] < b[1] if a[0] == b[0] else a[0] < b[0]
    else:
        return a[1] > b[1] if a[0] == b[0] else a[0] > b[0]


N, M = map(int, input().split())

points = []
for j in range(N):
    x1, x2 = map(int, input().split())
    points.append((x1, 0))
    points.append((x2, 2))

P = list(map(int, input().split()))
for i in range(M):
    points.append((P[i], 1, i))

points = quickSort(points)

count = 0
for k in range(len(points)):
    point = points[k]
    if (point[1] == 1):
        P[point[2]] = str(count)
    if (point[1] == 0):
        count += 1
    if (point[1] == 2):
        count -= 1

print(" ".join(P))


current, peak = tracemalloc.get_traced_memory()
print(
    f"Current memory usage is {current / 10**6}MB; Peak was {peak / 10**6}MB")
tracemalloc.stop()
